text/trace/0219-payload-collection.md

```diff
@@ -0,0 +1,341 @@
+functionality to trace API, and to OTLP. As we show in this proposal, adding such data using
+the current APIs is limited and problematic.
+
+What do we mean by payload data? While it’s hard to precisely define, the general guidelines are:
+
+- It includes the content of a message, rather than metadata or headers
+- It includes information from the “data plane”, rather than the “control plane”
+
+Some example of payload data includes HTTP bodies, database queries (including items read or written), and messages produced or consumed from a queue.
+
+The possible value from collecting payloads is substantial. Using payload data, OpenTelemetry users can troubleshoot applications much more effectively in many cases - they can use that to understand exact data flows in their systems, reproduce problematic requests, or search for traces using specific payload information while troubleshooting.
+
+## Explanation
+
+Most generally, payload data is just a binary buffer. Mostly it will have a defined encoding, used by the code pieces which work with it to decode its content. For modern applications, standard encoding is usually used. For strings, UTF-8 or ASCII is most common,
+while other encodings are usually used to represent nested mappings, like JSON, YAML, Protobuf, Avro, and many more.
+
+While collecting binary buffers may be useful in some cases, collecting decoded
+objects could be much more useful as processors and backends will be able to access internal
+fields in a standard and performant way. We also show other capabilities that could be
+accomplished by using a standard encoding for payload data.
+
+## Internal details
+
+We propose adding a new type of span attributes, called '**payload attributes**', intended for storing decoded payload data. This will be implemented by additions of new fields and data types to Span proto definition, and API methods to support it. SDKs and OTel collector will be updated to support these changes as well.
+
+The API & functionality of current Span attributes will remain the same, as they will
+still be used for collecting general-purpose, non-payload data. Therefore, the proposed changes are **non-breaking**. The only potential breaking change is regarding certain semantic conventions which may fit better as payload attributes, such as `db.statement`.
+
+### OTLP Updates
+
+We describe a prototype of additions to OTLP, to support encoding JSON-like objects in spans, together with some
+extra metadata regarding the original plain payload. This prototype is likely to be changed during specifications and formats discussions
+but hopefully could set basic characteristics.
+
+We propose using a similar structure to the native protobuf [Struct](https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/struct.proto#L51) message,
+(which is a general representation of a JSON object), with some embedded metadata -
+
+```protobuf
+message Value {
```

Comment:
can you not use the existing AnyValue?

https://github.com/open-telemetry/opentelemetry-proto/blob/724e427879e3d2bae2edc0218fff06e37b9eb46e/opentelemetry/proto/common/v1/common.proto#L28

Reply:
There are a few important differences - supporting `NullValue` and the 'nested' metadata fields (and not supporting `bytes`)

Reply:
How is NullValue different from the "empty" value that AnyValue supports? What 'nested' metadata fields are you referring to? Do you mean `original_encoding` and others?

Reply:
* That's right about "empty" value (missed it)
* You are correct about the metadata fields